### PR TITLE
Minor updates for Firestore

### PR DESCRIPTION
**Pull request by jskeet**
(As requested by the Firestore team)

Best reviewed one commit at a time.

### CODE REVIEW

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Precondition.cs [hunk 1]
```diff
@@ -20,7 +20,7 @@ namespace Google.Cloud.Firestore
     /// <summary>
     /// Immutable class representing a precondition for an update operation.
     /// </summary>
-    public sealed class Precondition : IEquatable<Precondition>
+    public sealed class Precondition
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Precondition.cs]: This isn't strictly an Options class. Just checking you definitely meant to remove Equality from here?

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Query.cs [hunk 2]
```diff
@@ -56,9 +56,9 @@ public class Query : IEquatable<Query>
         // Parent path of this query
         internal string ParentPath => Collection.Parent?.Path ?? Database.DocumentsPath;
 
-        // This would be protected, but that would allow subclasses from other assemblies. The intention is that the only concrete
-        // subclass of Query is CollectionReference. If "private protected" ever ends up in C#, this constructor can be changed.
-        internal Query()
+        // private protected = only visible to subclasses within the same assembly.
+        // In reality, *only* CollectionReference will ever derive from Query.
+        private protected Query()
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Query.cs]: Living on the edge :)

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Precondition.cs (hunk 1 above)
Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Precondition.cs]: Yup, definitely.

**Comment by chrisdunelm** [comment]
One last comment seemed to get lost...

Should there be tests for the new `Update[Async]` methods on `Documentreference` and `Transaction`?

**Comment by jskeet** [comment]
There aren't unit tests for the original methods... I could add integration tests I guess. I'm not overly worried about them, given that they're simple delegation.

**Comment by chrisdunelm** [comment]
OK, LGTM

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/DocumentReference.cs [hunk 3]
```diff
@@ -119,6 +119,33 @@ public CollectionReference Collection(string path)
...
+        /// <param name="cancellationToken">A cancellation token to monitor for the asynchronous operation.</param>
+        /// <returns>The write result of the server operation.</returns>
+        public Task<WriteResult> UpdateAsync(IDictionary<string, object> updates, Precondition precondition = null, CancellationToken cancellationToken = default)
+        {
+            GaxPreconditions.CheckNotNull(updates, nameof(updates));
+            return UpdateAsync(updates.ToDictionary(pair => FieldPath.FromDotSeparatedString(pair.Key), pair => pair.Value), precondition, cancellationToken);
+        }
+
+        /// <summary>
+        /// Asynchronously performs a set of updates on the document referred to by this path, with an optional precondition.
```
Comment (evildour) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/DocumentReference.cs]: "a set of updates" => "an update"?

Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/DocumentReference.cs]: Humbug - thanks, will fix. (It's probably wrong in Transaction too.)

Comment (chrisdunelm) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/DocumentReference.cs]: Well spotted ;)

**Pull request merged**